### PR TITLE
Fix distributed native vision loading

### PR DESCRIPTION
**Pull request by ttupper92618**
## Summary

- load distributed MLX vision shards through the native MLX-VLM loader when a model card's vision weights are the model's primary checkpoint
- share the loader-selection rule with the existing single-node path
- add a focused regression test proving pipeline shards retain the native VLM loader contract

## Root cause

MLX-LM recognizes some Qwen VLM checkpoints as text models and strips their vision towers. Single-node placements already bypassed that behavior, but distributed `shard_and_load` did not. Multi-rank requests therefore fell back to precomputed image embeddings without the model family's image-grid-aware positions and produced deterministic OCR corruption while still recognizing coarse color and shape attributes.

## User impact

Distributed Apple vision placements now preserve the same native multimodal model contract as single-node placements instead of producing plausible but incorrect image text.

## Validation

- `uv run basedpyright`
- `uv run ruff check`
- `nix --extra-experimental-features 'nix-command flakes' fmt`
- `uv run pytest` — 2932 passed, 1 skipped, 228 deselected
- live two-node Apple qualification using an image-only hidden code fixture — 2/2 exact code, color, and shape assertions passed

### CODE REVIEW

File: src/skulk/worker/engines/mlx/utils_mlx.py [hunk 1]
```diff
@@ -1071,7 +1081,12 @@ def shard_and_load(
         shard_metadata.model_card.source_revision,
     )
 
-    model, _ = load_model(model_path, lazy=True, strict=False)
+    model, _ = load_model(
+        model_path,
+        lazy=True,
+        strict=False,
+        prefer_vlm=_prefers_native_vlm(shard_metadata.model_card),
```
[the remark this answers is not in the document]
Comment (ttupper92618) [reply src/skulk/worker/engines/mlx/utils_mlx.py:1106]: Severity 5 (likelihood 5, impact 5, triggerability 0, complexity 0; score 5.0). Fixed: native VLM tensor placement delegates sharding to the nested language trunk while keeping the logits-returning wrapper as the collective dependency target, and runtime identities include the actual qwen3_5_text/qwen3_5_moe_text values. Full validation passed (2,936 tests, 1 skipped, strict typing/lint/format clean); a live two-node Tensor placement became ready and passed the exact vision fixture twice with zero issues.

File: src/skulk/worker/engines/mlx/auto_parallel.py [hunk 2]
```diff
@@ -840,16 +850,33 @@ def patched_call(
...
-            if hasattr(dep_cache, "keys"):  # type: ignore
-                dep_cache.keys = mx.depends(dep_cache.keys, logits)  # pyright: ignore[reportAny]
+        if isinstance(cache, (list, tuple)) and cache:
+            last = cast(object, cache[-1])
+            dep_cache = (
+                cast(_CacheContainer, last)[0] if hasattr(last, "caches") else last
+            )
+            if hasattr(dep_cache, "keys"):
+                cache_with_keys = cast(_CacheWithKeys, dep_cache)
+                cache_with_keys.keys = mx.depends(cache_with_keys.keys, logits)
```
[the remark this answers is not in the document]
Comment (ttupper92618) [reply src/skulk/worker/engines/mlx/auto_parallel.py:881]: Severity 2 (likelihood 2, impact 4, triggerability 1, complexity 1; score about 2.3). No change in this PR. The dependency patch runs after the original forward; production cache factories return lists, and a final KV entry has populated keys after that forward (non-KV entries do not expose keys). The live first-request Tensor proof passed twice. A generic cold-cache guard can be handled separately if a production cache constructor is observed to retain a null key slot.

File: src/skulk/worker/engines/mlx/utils_mlx.py [hunk 3]
```diff
@@ -1071,7 +1099,12 @@ def shard_and_load(
...
```
[the remark this answers is not in the document]
Comment (ttupper92618) [reply src/skulk/worker/engines/mlx/utils_mlx.py:1106]: Severity 4 (likelihood 4, impact 5, triggerability 0, complexity 1; score 4.3). Fixed in 1105fd75: native Qwen inner trunks now enter the same hybrid-cache reindexing path as MLX-LM trunks, including local fa_idx/ssm_idx recomputation after slicing. Added a non-aligned three-rank regression. Full validation passed, and a live three-node Pipeline placement became ready and passed the exact vision fixture twice with zero issues.

File: src/skulk/worker/engines/mlx/auto_parallel.py [hunk 4]
```diff
@@ -840,16 +850,33 @@ def patched_call(
...
+            )
 
         # Add dependency to last cache entry to ensure distributed ops are evaluated
-        if cache is not None and len(cache) > 0:  # pyright: ignore[reportAny]
-            last = cache[-1]  # pyright: ignore[reportAny]
-            dep_cache = last[0] if hasattr(last, "caches") else last  # pyright: ignore[reportAny]
-            if hasattr(dep_cache, "keys"):  # type: ignore
-                dep_cache.keys = mx.depends(dep_cache.keys, logits)  # pyright: ignore[reportAny]
+        if isinstance(cache, (list, tuple)) and cache:
+            last = cast(object, cache[-1])
```
[the remark this answers is not in the document]
Comment (ttupper92618) [reply src/skulk/worker/engines/mlx/auto_parallel.py:875]: Severity 1 (likelihood 1, impact 4, triggerability 2, complexity 2; score about 1.4). No change in this PR. KVCacheType is intentionally broad at the type boundary, but every production cache factory on this path returns a list; no non-list Sequence constructor exists in the supported MLX/MLX-VLM families. The two-node Tensor live proof exercised the production cache shape successfully.

File: src/skulk/worker/engines/mlx/utils_mlx.py [hunk 5]
```diff
@@ -740,6 +758,20 @@ def load_model(
...
+
+    MLX-LM recognizes some multimodal checkpoints as text models and strips
+    their vision towers. When the card points vision at the same checkpoint as
+    its language weights, every placement shape must load through MLX-VLM so
+    family-specific image-grid positions and embeddings remain available.
+    """
+    vision_config = model_card.vision
+    return vision_config is not None and vision_config.weights_repo == str(
+        model_card.model_id
+    )
```
[the remark this answers is not in the document]
Comment (ttupper92618) [reply src/skulk/worker/engines/mlx/utils_mlx.py:772]: Severity 1 (likelihood 1, impact 4, triggerability 1, complexity 1; score about 1.6). No change in this PR. shard_and_load receives a validated ModelCard, whose after-validator replaces an empty bundled vision weights_repo with the primary model ID. Tests that construct VisionCardConfig() inside ModelCard exercise that normalization, so this branch cannot observe the empty value described here.
